sdk/typespec/typespec_client_core/src/http/pager.rs

```diff
@@ -0,0 +1,256 @@
+    /// The `C` type parameter is the type of the continuation. It may be any [`Send`]able type.
+    /// The result will be an asynchronous stream of [`Result<Response<T>>`](typespec::Result<Response<T>>) values.
+    ///
+    /// The first time your callback is called, it will be called with [`Option::None`], indicating no continuation value is present.
+    /// Your callback must return one of:
+    /// * `Ok((response, Some(continuation)))` - The request succeeded, and return a response `response` and a continuation value `continuation`. The response will be yielded to the stream and the callback will be called again immediately with `Some(continuation)`.
+    /// * `Ok((response, None))` - The request succeeded, and there are no more pages. The response will be yielded to the stream, the stream will end, and the callback will not be called again.
+    /// * `Err(..)` - The request failed. The error will be yielded to the stream, the stream will end, and the callback will not be called again.
+    ///
+    /// ## Examples
+    ///
+    /// ```rust,no_run
+    /// # use typespec_client_core::http::{Context, Pager, Pipeline, Request, Response, Method, headers::HeaderName};
+    /// # let pipeline: Pipeline = panic!("Not a runnable example");
+    /// # struct MyModel;
+    /// let url = "https://example.com/my_paginated_api".parse().unwrap();
+    /// let mut base_req = Request::new(url, Method::Get);
+    /// let pager = Pager::from_callback(move |continuation| {
+    ///     // The callback must be 'static, so you have to clone and move any values you want to use.
+    ///     let pipeline = pipeline.clone();
+    ///     let mut req = base_req.clone();
+    ///     async move {
+    ///         if let Some(continuation) = continuation {
+    ///             req.insert_header("x-continuation", continuation);
+    ///         }
+    ///         let resp: Response<MyModel> = pipeline
+    ///           .send(&Context::new(), &mut req)
+    ///           .await?;
+    ///         let continuation_token = resp.headers().get_optional_string(&HeaderName::from_static("x-next-continuation"));
+    ///         Ok((resp, continuation_token))
+    ///     }
+    /// });
+    /// ```
+    pub fn from_callback<
+        // This is a bit gnarly, but the only thing that differs between the WASM/non-WASM configs is the presence of Send bounds.
+        #[cfg(not(target_arch = "wasm32"))] C: Send + 'static,
+        #[cfg(target_arch = "wasm32")] C: 'static,
+        #[cfg(not(target_arch = "wasm32"))] F: Fn(Option<C>) -> Fut + Send + 'static,
+        #[cfg(target_arch = "wasm32")] F: Fn(Option<C>) -> Fut + 'static,
+        #[cfg(not(target_arch = "wasm32"))] Fut: Future<Output = Result<(Response<T>, Option<C>), typespec::Error>> + Send + 'static,
+        #[cfg(target_arch = "wasm32")] Fut: Future<Output = Result<(Response<T>, Option<C>), typespec::Error>> + 'static,
```

Comment:
Could we use `cfg_if` instead? It would halve the number of lines for improved readability.

Reply:
Ooh, possibly! I'll look in to that

Reply:
Alas, no. You can't invoke a macro within a generic arguments block or in where clauses. There are basically four ways we can do this:

1. Two completely separate `from_callback` functions, with duplicated bodies, marked with `#[cfg]` attributes and appropriate type bounds
2. This layout, with `#[cfg]` attributes on the trait bounds themselves.
3. Wrapping the entire declaration in a macro, invoked separately for each `#[cfg]` variant (this is what the old code did)
4. Possibly: A custom attribute macro we could write (I didn't see one on crates.io) to do this for us (something like `#[bound_if(not(target_arch = "wasm32"), Send)] C: 'static`)

I wasn't a fan of 3 in the original since it made error messages and F12 go-to-definition a little wonky, but I could live with it. What I like about option 2 is that while it's a little gnarly, it's fairly clear what's going on and you can largely ignore it. I did rearrange it a bit to this:

```rust
    pub fn from_callback<
        // This is a bit gnarly, but the only thing that differs between the WASM/non-WASM configs is the presence of Send bounds.
        #[cfg(not(target_arch = "wasm32"))] C: Send + 'static,
        #[cfg(not(target_arch = "wasm32"))] F: Fn(Option<C>) -> Fut + Send + 'static,
        #[cfg(not(target_arch = "wasm32"))] Fut: Future<Output = Result<(Response<T>, Option<C>), typespec::Error>> + Send + 'static,

        #[cfg(target_arch = "wasm32")] C: 'static,
        #[cfg(target_arch = "wasm32")] F: Fn(Option<C>) -> Fut + 'static,
        #[cfg(target_arch = "wasm32")] Fut: Future<Output = Result<(Response<T>, Option<C>), typespec::Error>> + 'static,
    >(
        make_request: F,
    ) -> Self {
```

Reply:
That layout had crossed my mind as well, but off hand wasn't sure how it'd look. Actually looks nice - more readable, certainly.